src/jit/flowgraph.cpp

```diff
@@ -8241,90 +8241,381 @@ void Compiler::fgAddInternal()
+                            constReturnBlock = CreateReturnBB(searchLimit, retConst);
+                        }
+                    }
+
+                    if (constReturnBlock != nullptr)
+                    {
+                        // Found a constant merged return block.
+                        mergedReturnBlock = constReturnBlock;
+
+                        // Change BBJ_RETURN to BBJ_ALWAYS targeting const return block.
+                        assert((comp->info.compFlags & CORINFO_FLG_SYNCH) == 0);
+                        returnBlock->bbJumpKind = BBJ_ALWAYS;
+                        returnBlock->bbJumpDest = constReturnBlock;
+
+                        // Remove GT_RETURN since constReturnBlock returns the constant.
+                        assert(returnBlock->lastStmt()->gtStmtExpr->OperIs(GT_RETURN));
+                        assert(returnBlock->lastStmt()->gtStmtExpr->gtGetOp1()->IsIntegralConst());
+                        comp->fgRemoveStmt(returnBlock, returnBlock->lastStmt());
+                    }
+                }
+            }
+
+            if (mergedReturnBlock == nullptr)
+            {
+                // No constant return block for this return; use the general one.
+                mergedReturnBlock = comp->genReturnBB;
+                if (mergedReturnBlock == nullptr)
+                {
+                    // No general merged return for this function yet; create one.
+                    // There had better still be room left in the array.
+                    assert(searchLimit < maxReturns);
+                    mergedReturnBlock = CreateReturnBB(searchLimit);
+                    comp->genReturnBB = mergedReturnBlock;
+                }
+            }
+
+            if (returnBlock != nullptr)
+            {
+                // Propagate profile weight and related annotations to the merged block.
+                // Return weight should never exceed entry weight, so cap it to avoid nonsensical
+                // hot returns in synthetic profile settings.
+                mergedReturnBlock->bbWeight =
+                    min(mergedReturnBlock->bbWeight + returnBlock->bbWeight, comp->fgFirstBB->bbWeight);
+                if (!returnBlock->hasProfileWeight())
+                {
+                    mergedReturnBlock->bbFlags &= ~BBF_PROF_WEIGHT;
+                }
+                if (mergedReturnBlock->bbWeight > 0)
+                {
+                    mergedReturnBlock->bbFlags &= ~BBF_RUN_RARELY;
+                }
+
+                // We don't actually rewrite the return here for the non-const cases; morph
+                // will handle that after some key transformations like rewriting tail calls
+                // and calls that return to hidden buffers.
+                // We do go ahead and decrement fgReturnCount to reflect that this block will
+                // eventually be rewritten.
+                comp->fgReturnCount--;
```

Comment:
I'm probably missing something here, but the comment seems stale. 

If we found a constant return for this return point then we did rewrite the return. Maybe the first part should say "we didn't rewrite the return for cases where the return will use the common return block"?

And the second part, seems like merge always "succeeds" unless it is in the eager create mode and so decrementing is something that is always done, whether or not rewriting happened.

Reply:
I mentioned "for the non-const cases" in the first clause, but yeah the way the method flows now there's not a convenient place for a codepath representing that distinction, so I've moved the comments about this up to the method headers.

Reply:
Maybe it's a parsing ambiguity -- is a "non-const" case describing the input or the output? I've been thinking it applied to the input....